### PR TITLE
nixos/switch-to-configuration: restart changed socket units

### PR DESCRIPTION
**Pull request by lheckemann**
###### Motivation for this change
The bugfix part of #50340

###### Things done

- [x] Tested via nixos/tests/mpd.nix
- [x] Tested via nixos/tests/systemd-confinement.nix
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc author @fpletz

### CODE REVIEW

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 1]
```diff
@@ -214,7 +214,17 @@ sub fingerprintUnit {
                 # Reload the changed mount unit to force a remount.
                 $unitsToReload{$unit} = 1;
                 recordUnit($reloadListFile, $unit);
-            } elsif ($unit =~ /\.socket$/ || $unit =~ /\.path$/ || $unit =~ /\.slice$/) {
+            } elsif ($unit =~ /\.socket$/) {
+                my $unitInfo = parseUnit($newUnitFile);
+                # If a socket unit has been changed, the corresponding
+                # service unit has to be stopped before the socket can
+                # be restarted.
```
Comment (domenkozar) [inline nixos/modules/system/activation/switch-to-configuration.pl]:
```suggestion
                # be restarted. Service unit will be started automatically by the socket.
```

**Review by flokli — COMMENTED** [review]
Can we also extend this to `.socket` for user sessions?

This should help with https://github.com/NixOS/nixpkgs/pull/71095

**Comment by domenkozar** [comment]
It's not clear to me how this relates to #71095, but I think there's no need to extend scope. The fix already improves socket handling, any further improvements can be done with whoever does them later on :)

**Comment by flokli** [comment]
This question was more of a in-context heads-up. Of course doesn't need to stall this PR :-)

#71095 changed .socket files for the systemd user session, and I assume those were not properly reloaded/restarted, causing some already running gpg-agents not finding pinentry anymore.

**Comment by primeos** [comment]
Unfortunately this causes severe regressions by restarting systemd services that should not be restarted, e.g. `dbus.service` (I assume due to `Requires=dbus.socket`). This does e.g. crash Sway (and probably other graphical sessions as well) after `nixos-rebuild switch`.

See #74626 for more details.

**Comment by domenkozar** [comment]
Reverted in master until that's fixed.

**Pull request merged**